### PR TITLE
Add --flush option to flush all output

### PR DESCRIPTION
**Pull request by karkhaz**

### CODE REVIEW

**Review by smowton — APPROVED** [review]
Do as cpplint asks; otherwise seems like a handy addition.

File: src/util/cout_message.h [hunk 1]
```diff
@@ -35,6 +35,18 @@ class console_message_handlert:public ui_message_handlert
...
 
   virtual void flush(unsigned level) override;
+
+  console_message_handlert()
+    : always_flush(false)
+  {
+  }
+
+  console_message_handlert(bool always_flush)
+    : always_flush(always_flush)
```
Comment (kroening) [inline src/util/cout_message.h]: Needs an 'explicit'

**Comment by kroening** [comment]
See comment for missing 'explicit'

**Review by martin-cs — APPROVED** [review]
Looks good.

**Review by martin-cs — CHANGES_REQUESTED** [review]
Sorry -- can we have the flush option and help string as macros and in the relevant header?  (Your timer patch does this well).

**Comment by martin-cs** [comment]
@karkhaz : still interested, from my point of view it just needs that change then it is good to go.

**Comment by karkhaz** [comment]
Thanks for your patience @kroening @martin-cs @tautschnig, I've rebased & addressed all comments on this; all tests are passing.

File: src/cbmc/cbmc_parse_options.h [hunk 2]
```diff
@@ -71,6 +71,7 @@ class optionst;
   "(arrays-uf-always)(arrays-uf-never)" \
   "(string-abstraction)(no-arch)(arch):" \
   "(round-to-nearest)(round-to-plus-inf)(round-to-minus-inf)(round-to-zero)" \
+  "(flush)" \
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.h]: That should be `OPT_FLUSH`  as is  used elsewhere.

File: src/util/cout_message.cpp [hunk 3]
```diff
@@ -88,12 +88,7 @@ void console_message_handlert::flush(unsigned level)
   // We flush after messages of level 6 or lower.
   // We don't for messages of level 7 or higher to improve performance,
   // in particular when writing to NFS.
-  if(level>=4)
-  {
-    if(level<=6)
-      std::cout << std::flush;
-  }
-  else
+  if(always_flush || level <= 6)
```
Comment (tautschnig) [inline src/util/cout_message.cpp]: That does not seem to be correct: the previous code would (intentionally) flush `std::cout` in some levels as that's where output is sent to.

File: src/util/ui_message.cpp [hunk 4]
```diff
@@ -130,11 +134,13 @@ void ui_message_handlert::print(
     {
     case uit::PLAIN:
     {
-      console_message_handlert console_message_handler;
+      console_message_handlert console_message_handler(always_flush);
```
Comment (tautschnig) [inline src/util/ui_message.cpp:137]: JSON and XML streams should also be flushed.

**Review by tautschnig — CHANGES_REQUESTED** [review]

**Comment by karkhaz** [comment]
@tautschnig that's done, all checks have passed. Would you mind checking that the JSON & XML flush is in the right place? I tested it by replacing the call to flush with `out << "foo\n"` and it seemed to do the right thing when I passed `--flush`, but not sure if it covers all cases.

File: src/cbmc/cbmc_parse_options.cpp [hunk 5]
```diff
@@ -990,6 +990,7 @@ void cbmc_parse_optionst::help()
     " --xml-interface              bi-directional XML interface\n"
     " --json-ui                    use JSON-formatted output\n"
     HELP_GOTO_TRACE
+    HELP_FLUSH
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp:993]: Could you please also add this to goto-analyzer (and possibly goto-diff)?

File: src/util/ui_message.cpp [hunk 6]
```diff
@@ -144,6 +150,8 @@ void ui_message_handlert::print(
       source_locationt location;
       location.make_nil();
       print(level, message, -1, location);
+      if(always_flush)
+        flush(1);
```
Comment (tautschnig) [inline src/util/ui_message.cpp]: Replace `1` by `level`

File: src/util/ui_message.cpp [hunk 7]
```diff
@@ -274,6 +282,8 @@ void ui_message_handlert::xml_ui_msg(
 
   out << result;
   out << '\n';
+  if(always_flush)
+    flush(1);
```
Comment (tautschnig) [inline src/util/ui_message.cpp]: Replace `1` by `level`

**Review by tautschnig — CHANGES_REQUESTED** [review]

**Comment by karkhaz** [comment]
Comments addressed. All tests are passing, except that the one that failed because Travis ran out of disk space. Should I re-force-push, or will somebody investigate this?

`level` is not in scope for `xml_ui_msg()`, so I added a `level` parameter to both that function and its caller `ui_msg()`. On the other hand, `json_ui_msg` does not need a `level` parameter since the caller flushes it.

**Review by tautschnig — APPROVED** [review]

**Comment by tautschnig** [comment]
> `level` is not in scope for `xml_ui_msg()`, so I added a `level` parameter to both that function and its caller `ui_msg()`. On the other hand, `json_ui_msg` does not need a `level` parameter since the caller flushes it.

Very true - but then wouldn't the same actually hold for `xml_ui_msg`?

(I have restarted the Travis job.)

**Comment by tautschnig** [comment]
@martin-cs If/when you have a minute, could you please re-review?

**Comment by karkhaz** [comment]
Yes, you're right, I didn't follow the control flow up far enough. it looks like flushing for both XML and JSON should be taken care of on line 153. I've just re-pushed.

**Comment by karkhaz** [comment]
@martin-cs this is ready to go, are you happy with it?

**Review by martin-cs — APPROVED** [review]
Looks fine to me.  I'm unsure about @tautschnig 's comment about JSON and XML streams.  I'd be tempted to say they should be flushed at the end of objects...

**Pull request merged**